### ISSUE TITLE
Cognito Passwordless

### ISSUE THREAD

**Issue by jamilsaadeh97** [aws-amplify/amplify-flutter#5728]
Hello,

I want to check if the new [Cognito passwordless](https://aws.amazon.com/blogs/aws/improve-your-app-authentication-workflow-with-new-amazon-cognito-features/) workflow will be supported in Amplify Flutter.
I saw that Amplify Swift and Android released this feature and I'm interested in adding this workflow to my apps. If yes, could you provide an estimate if it's planned to be released soon?

Thank you!

**Comment by ekjotmultani** [comment]
Hi @jamilsaadeh97, thanks for your interest in the Cognito passwordless workflow for Amplify Flutter! We are looking into integrating this feature as well, however we can't provide any timelines yet. Please let me know if you have any other questions I can help with!

**Comment by jamilsaadeh97** [comment]
Hi @ekjotmultani 
Thank you for the update!

**Comment by hcocpinell** [comment]
Until this feature is released, is there any workaround to be able to start a SSO flow, call new Cognito UI and then go back with the returned credentials and set them on Amplify Flutter?

Response from cognito:

https://www.example.com#id_token=eyJra67890EXAMPLE&access_token=eyJra12345EXAMPLE&token_type=bearer&expires_in=3600&state=abcdefg

Manually extract the credentials from javascript and pass them to Flutter:

amplify.someMethodToSetCredentials(idToken, accessToken)

Is this possible?

**Comment by davidcaseria** [comment]
In the meantime, are any workarounds available to support passwordless sign-in using the existing APIs? Thanks!

**Comment by tyllark** [comment]
Hello @davidcaseria, we don't have any work around currently for using passwordless with existing APIs. We don't have a timeline for passwordless implementation yet, but the other Amplify libraries (JavaScript, Android, iOS) recently released this feature. We will post an update here once the feature is released!

**Comment by amankachhal99** [comment]
Is this released yet ?

**Comment by harsh62** [comment]
@amankachhal99 The team is currently working on supporting this on Flutter, we will post an update on the issue once this is done.

**Comment by amankachhal99** [comment]
@harsh62 .. thanks for reply. 

Any timeline for release? Insight in this week help us to plan our tasks.

**Comment by harsh62** [comment]
@amankachhal99 We don't have a timeline yet, but we are actively developing the feature. _May be_ something end of next month. 

**NOTE**: Please keep in mind that this timeline could change and is not a commitment for delivery.

**Comment by amankachhal99** [comment]
Hi @tyllark, @harsh62 
I see that this PR is merged.. I understand you guys must be having your own feature priority list.
But i just wanted to request if this feature release can be prioritised. We are pending for this feature and lot of big features in my app are blocked because of this.

Just wanted to see if there is any way to bring more attention to this.. :)

Cheers to all the amazing work you guys are doing.. 👍

**Comment by jhw** [comment]
Just adding 👍 that I would love to see this ASAP

**Comment by mattcreaser** [comment]
Hi @jhw and @amankachhal99. @tyllark is working hard on the passwordless support. We are going to release the first part (passwordless with email and sms OTP) as soon as it's ready, and then follow up with webAuthN support afterwards. I do not have a date but rest assured it is a priority.

**Comment by hrmanika-blip** [comment]
Adding comment too as it's urgently required in my project too and all other amplify sdk already have it.  Sad too not see in flutter amplify.  
Please help to bring it soon.  🫰

**Comment by gauravsenpai23** [comment]
Hi team,

This feature is really important for our project, and we’ve been waiting for it on an urgent basis. Our team is kind of stuck without it, so it would be awesome if you could release it ASAP.

Thanks a lot for all the great work you’re doing really looking forward to seeing this one go live!

**Comment by amankachhal99** [comment]
Hi @mattcreaser  , @harsh62  @tyllark ,
As we approach the end of the month, I wanted to kindly check if there is a concrete timeline for the release of this feature. Many people are eagerly awaiting it, so any update would be greatly appreciated.

Thank you and best regards,

**Comment by tyllark** [comment]
Hello @amankachhal99, we do not have a concrete date set yet. We paused development on the feature for a couple weeks to to focus on operation issues, but Passwordless is back to being my primary focus now. Once phase 1 of the feature (passwordless with email and sms OTP without Authenticator) is being reviewed internally I will provide you with a pubspec override for the feature branch, so you can get a head start on updating your app.

**Comment by amankachhal99** [comment]
Hi @tyllark , 
Any update on the feature ? it has already been months for the feature request. 
I feel flutter library is already way behind other language libraries. :(

**Comment by tyllark** [comment]
Hello @amankachhal99, thank you for your patients I understand the delay has been frustrating. I'm wrapping up documentation change and waiting for final sign off before we can create the final Pull Requests. Ill provide the pubspec override I promised later today, so you can start updating your app.

**Comment by amankachhal99** [comment]
Ok Waiting.

**Comment by tyllark** [comment]
Hello @amankachhal99, Here are the steps for how you can test the Passwordless changes in your app:

1. Enable Passwordless in Cognito, I did this by updating my `amplify/backend.ts` file as follows:
```
import { defineBackend } from '@aws-amplify/backend';
import { auth } from './auth/resource';

const backend = defineBackend({
  auth,
});

const { cfnResources } = backend.auth.resources;
const { cfnUserPool, cfnUserPoolClient } = cfnResources;

// enable ASF
cfnUserPool.userPoolAddOns = {
  advancedSecurityMode: "AUDIT",
}

// Specify which authentication factors you want to allow with USER_AUTH
cfnUserPool.addPropertyOverride(
	'Policies.SignInPolicy.AllowedFirstAuthFactors',
	['PASSWORD', 'EMAIL_OTP', 'SMS_OTP']
);

// The USER_AUTH flow is used for passwordless sign in
cfnUserPoolClient.explicitAuthFlows = [
	'ALLOW_REFRESH_TOKEN_AUTH',
	'ALLOW_USER_AUTH'
];
```
2. Clone the Amplify Flutter repo and bootstrap it as defined in our [CONTRIBUTING.md](https://github.com/aws-amplify/amplify-flutter/blob/main/CONTRIBUTING.md#setting-up-for-local-development).
3. Check out the `feat/passwordless-preview` branch.
4. Override your app's dependency with your local copy of Amplify Flutter by adding the attached [pubspec_overrides.yaml](https://github.com/user-attachments/files/22617629/pubspec_overrides.yaml) to your projects root folder. Please replace the `[PATH_TO_REPO]` placeholder with the path you clone Amplify Flutter.
5. Test your app out

I'm able to Signup without a password:
```
final signUpResult = await Amplify.Auth.signUp(
  username: username, //My backend is configured to use emails for usernames
);

final confirmSignUpResult = await Amplify.Auth.confirmSignUp(
  username: username,
  confirmationCode: 'Code Received In Email',
);
```

and sign in without a password:
```
final signInResult = await Amplify.Auth.signIn(
  username: username,
  options: const SignInOptions(
    pluginOptions: CognitoSignInPluginOptions(
      authFlowType: AuthenticationFlowType.userAuth,
    ),
  ),
);

final confirmFirstFactorResult = await Amplify.Auth.confirmSignIn(
  confirmationValue: AuthFactorType.emailOtp.value, //signInResult.nextStep.availableFactors lists the options
);

final confirmSignInResult = await Amplify.Auth.confirmSignIn(
  confirmationValue: 'Code Received In Email',
);
```
Please let me know if you have any questions. Here is the  [Amplify Android](https://docs.amplify.aws/android/build-a-backend/auth/concepts/passwordless/) documentation with additional examples. As a reminder WenAuthN is disabled to this phase of the relese.

**Comment by amankachhal99** [comment]
@tyllark just to double confirm that the feature works for both Android and iOS ?

Also, do we have any know issues as of now ?

**Comment by harsh62** [comment]
@amankachhal99 It should work on both iOS and Android.

**Comment by r0wdiggity** [comment]
I am getting the following error when trying to run the `password-preview` fork

```
Launching lib/main.dart on Linux in debug mode...
ERROR: No file or variants found for asset: packages/amplify_secure_storage_dart/lib/src/worker/workers.min.js.
ERROR: Target debug_bundle_linux-x64_assets failed: Exception: Failed to bundle asset files.
Building Linux application...                                           
Error: Build process failed
```

**Comment by tyllark** [comment]
Hello @r0wdiggity, are you getting any errors when you run `aft bootstrap` as per the [CONTRIBUTING.md](https://github.com/aws-amplify/amplify-flutter/blob/main/CONTRIBUTING.md#setting-up-for-local-development)?

This command will run [build_runner](https://pub.dev/packages/build_runner) on our packages which will trigger [build_web_compilers](https://pub.dev/packages/build_web_compilers) to generate the `workers.min.js` file and then our [WorkerCopyBuilder](https://github.com/aws-amplify/amplify-flutter/blob/main/packages/worker_bee/worker_bee_builder/lib/src/copy_builder.dart) should copy it to the correct asset position.

**Comment by amankachhal99** [comment]
When are we planning for release of this ?

**Comment by tyllark** [comment]
Hello @amankachhal99, we are waiting for final approvals from  a different team before we can merge the changes. I'll ask them for an ETA and get back to you, but once everything is approved we will merge and release.

**Comment by tyllark** [comment]
Hi @amankachhal99, I was given an estimate of ~2 weeks until the review is started. I'll provide an update here once the changes are merged in. Thank you for your patience, I know this feature has been delayed too many times and I apologize for the inconvenience.

**Comment by amankachhal99** [comment]
Hi @tyllark , 
Yes we have been pending a major user onboarding revamp just because of this feature.

Thanks for your follow ups and dev efforts.

**Comment by amankachhal99** [comment]
Hi @tyllark , 
It has already been 3 weeks, any expected release window visible ?

**Comment by tyllark** [comment]
Hi @amankachhal99, I'm actively following up with the review team. They're currently prioritizing re:Invent tasks, but I'm working to get a concrete timeline for when we can merge and release after that. I'll provide an update here as soon as I have more information. Thanks for your patience.

**Comment by amankachhal99** [comment]
Hi @tyllark , Greatings for the day..
This issue had already crossed 1 year. When are we expecting its release ?

**Comment by tyllark** [comment]
Hello @amankachhal99, I do not have an ETA for the release at this time. I'll provide an update here when I have additional information.

**Comment by amankachhal99** [comment]
Hi @tyllark , 
Its getting overwhelming to wait for an year for a feature which has been in market for so long on other frameworks. 

Is there any way I can sensitise this feature release and priority with the approval teams ? This has become a long dark tunnel.

**Comment by tyllark** [comment]
Hi @amankachhal99,
I had a meeting with the approval team and they assigned someone to review starting Jan 12th. I can't provide an exact ETA, but with the holiday season wrapping up and the re:invent crunch done I'm hopeful that there will not be anymore road blocks. Ideally we should have have prioritized such a major feature to release at the same time last year as Amplify JS/Android/Swift.

**Comment by amankachhal99** [comment]
@tyllark any updates? Its already 16th Jan

**Comment by cadivus** [comment]
Hi @amankachhal99,

We made some progress but can't provide an exact ETA yet.

**Comment by amankachhal99** [comment]
Hi @tyllark ,
I see that PR has been merged and I can also see a version published on pub. So can we consider that this feature has been released completely? Can we start using it?

**Comment by cadivus** [comment]
Hi @amankachhal99 and @jamilsaadeh97 ,
you can use the password feature now. WebAuthn is not supported yet but the other flows are.
For using it, please update the amplify-flutter lib version.

**Comment by ersantos1991** [comment]
Thanks for the last update. I’ve been testing it with the references for a few weeks, and everything is working well.

**Comment by jvh-aws** [comment]
@ersantos1991 that's good to hear!

**Comment by mraszplewicz** [comment]
Hi @tyllark @cadivus ,

Are you sure it works with SMS? I am getting `InvalidStateException` with `Unrecognized challenge type: SMS_OTP` message. I am just doing:
```
final result = await Amplify.Auth.signIn(
  username: phoneNumber,
  options: const SignInOptions(
    pluginOptions: CognitoSignInPluginOptions(
      authFlowType: AuthenticationFlowType.userAuth,
      preferredFirstFactor: AuthFactorType.smsOtp,
    ),
  ),
);
```

I can provide a full example if you need it.

**Comment by cadivus** [comment]
Hi @mraszplewicz,
thanks for reporting this. We can reproduce it and look into this.

**Comment by mraszplewicz** [comment]
I created a PR to fix SMS OTP: https://github.com/aws-amplify/amplify-flutter/pull/6632

**Comment by jamilsaadeh97** [comment]
Thank you for the feature @cadivus 
I have two questions:
- Can an already deployed amplify backend with username/password enable passwordless without recreating the userpool?
- Can we use a custom sms sender to send the otp codes with?

**Comment by cadivus** [comment]
Hi @jamilsaadeh97,
> Can an already deployed amplify backend with username/password enable passwordless without recreating the userpool?

Yes, that works.

> Can we use a custom sms sender to send the otp codes with?

Is [this](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sms-sender.html) what you are looking for (a completely custom sender)?  

Or this one (custom sender ID)?  

<img width="815" height="641" alt="Image" src="https://github.com/user-attachments/assets/76014fa3-849e-40aa-892d-e191112d648b" />

**Comment by jamilsaadeh97** [comment]
> Is [this](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sms-sender.html) what you are looking for (a completely custom sender)?

Yes using a Custom SMS Sender lambda instead of sending by default with SNS.
I already have implemented this lambda, just making sure with passwordless activated, it will still be used instead of SNS

**Comment by cadivus** [comment]
@jamilsaadeh97 Yes, this works.

**Comment by kishore-ac9** [comment]
Hello,
I wanted to follow up on this feature request regarding Cognito passwordless authentication in Amplify Flutter.
In my current Flutter project, we are using AWS Cognito for authentication. Since Amplify Flutter does not yet support passwordless workflows (such as email OTP or SMS OTP), we had to implement this functionality using native Cognito Identity Provider SDKs on both iOS and Android, and then bridge it back to Flutter.

Also, in our setup, we are not using any custom authentication flow—we haven’t configured custom auth challenges or added any Lambda triggers in Cognito. Our requirement is to use the standard USER_AUTH flow, not CUSTOM_AUTH.

While this workaround works, it introduces additional complexity in terms of:

- Maintaining platform-specific code
- Handling consistency between iOS and Android implementations
- Increasing overall development and testing effort

Having native support for passwordless authentication directly in Amplify Flutter (especially for USER_AUTH-based flows like email/SMS OTP) would significantly simplify the architecture and improve developer experience.

Could you please share if there have been any updates on this feature, Is it already available or currently on the roadmap?
Thanks again for your support!

**Comment by cadivus** [comment]
Hi @kishore-ac9,

Passwordless authentication with the `USER_AUTH` flow (including email OTP and SMS OTP) is supported in Amplify Flutter since **`amplify_auth_cognito` v2.10.0** - just our UI components are not updated yet.

### What's available today

You can use the standard `Amplify.Auth.signIn` and `Amplify.Auth.confirmSignIn` APIs with the `USER_AUTH` flow to implement passwordless sign-in using email OTP or SMS OTP — no need for native Cognito Identity Provider SDKs or platform-specific code.

### What's NOT yet available

The **pre-built Authenticator UI widget** (`amplify_authenticator`) does not yet have built-in support for the passwordless-specific steps like first factor selection (`continueSignInWithFirstFactorSelection`) or the `confirmSignInWithPassword` step. This is still under construction. For now, you'll need to build a custom UI to handle these steps.

### How to use it with custom UI

Here's the key API surface:

**1. Sign in with `USER_AUTH` flow:**
```dart
final result = await Amplify.Auth.signIn(
  username: 'user@example.com',
  options: const SignInOptions(
    pluginOptions: CognitoSignInPluginOptions(
      authFlowType: AuthenticationFlowType.userAuth,
      // Optionally hint the preferred first factor:
      preferredFirstFactor: AuthFactorType.smsOtp, // or emailOtp
    ),
  ),
);
```

**2. Handle the sign-in steps in a loop:**
```dart
switch (result.nextStep.signInStep) {
  case AuthSignInStep.continueSignInWithFirstFactorSelection:
    // Present available factors to the user
    final factors = result.nextStep.availableFactors;
    // e.g. {AuthFactorType.emailOtp, AuthFactorType.smsOtp, AuthFactorType.password}
    // User selects one, then:
    final confirmResult = await Amplify.Auth.confirmSignIn(
      confirmationValue: selectedFactor.value, // e.g. 'EMAIL_OTP'
    );
    break;

  case AuthSignInStep.confirmSignInWithOtpCode:
    // User received an OTP via email or SMS
    // Delivery details available at: result.nextStep.codeDeliveryDetails
    final otpResult = await Amplify.Auth.confirmSignIn(
      confirmationValue: otpCode,
    );
    break;

  case AuthSignInStep.confirmSignInWithPassword:
    // If user selected password as the factor
    final pwResult = await Amplify.Auth.confirmSignIn(
      confirmationValue: password,
    );
    break;

  case AuthSignInStep.done:
    // Sign-in complete!
    break;
}
```

**3. Available `AuthFactorType` values:**
- `AuthFactorType.emailOtp` — OTP sent to email
- `AuthFactorType.smsOtp` — OTP sent to SMS
- `AuthFactorType.password` — Standard password
- `AuthFactorType.passwordSrp` — Password via SRP
